pkg/block/fetcher.go

```diff
@@ -574,4 +581,61 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// IgnoreDeletionMarkFilter is a MetaFetcher that filters blocks that are marked for deletion; (delay/2) duration before current time.
+// The delay duration is to make sure that the replacement block can be fetched before we filter out the old block.
+type IgnoreDeletionMarkFilter struct {
+	logger           log.Logger
+	delay            time.Duration
+	bkt              objstore.BucketReader
+	deletionMarkIds  []ulid.ULID
+	toDeleteBlockIds []ulid.ULID
+}
+
+// NewIgnoreDeletionMarkFilter creates IgnoreDeletionMarkFilter.
+func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.BucketReader, delay time.Duration) *IgnoreDeletionMarkFilter {
+	return &IgnoreDeletionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+		delay:  delay,
+	}
+}
+
+// DeletionMarkBlocks returns block ids that have deletion-mark.json file.
+func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() []ulid.ULID {
+	return f.deletionMarkIds
+}
+
+// ToDeleteBlocks returns block ids that were marked for deletion; delay duration time ago.
+func (f *IgnoreDeletionMarkFilter) ToDeleteBlocks() []ulid.ULID {
+	return f.toDeleteBlockIds
+}
+
+// Filter filters out blocks that are marked for deletion; (delay/2) duration before current time.
+// It also returns the blocks that can be deleted since they were uploaded delay duration before current time.
+func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	f.deletionMarkIds = []ulid.ULID{}
+	f.toDeleteBlockIds = []ulid.ULID{}
+
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err != nil {
+			if err == metadata.ErrorDeletionMarkNotFound {
+				continue
+			}
+			return err
+		}
+		f.deletionMarkIds = append(f.deletionMarkIds, id)
+		if time.Since(time.Unix(deletionMark.DeletionTime, 0)).Seconds() > (f.delay.Seconds())/2 {
+			synced.WithLabelValues(markedForDeletionMeta).Inc()
+			delete(metas, id)
+		}
+		if time.Since(time.Unix(deletionMark.DeletionTime, 0)).Seconds() > f.delay.Seconds() {
```

Comment:
What about having just list of metas WITH deletion json and allow cleaner to decide about that? It will keep this filter focused and not confusing for other readers who will use this filter. (:

Reply:
This is really very clean. Thanks for the feedback. working on it

Reply:
I have made this change now. It looks cleaner now